### ISSUE TITLE
Improve implementation of QualifiedName.hashCode()

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#13499]
A profiling session produced a hotspot implementation of `QualifiedName::hashCode`

![image](https://user-images.githubusercontent.com/734593/165483404-fcc7ae69-84ca-466c-ae49-a965236bfdd9.png)

While such hotspots can often be profiling artifacts, in this case, it is safe to say that:

1. Calling `QualifiedName::getName` is inefficient, with its iteration and `String[]` allocation
2. `Name` are immutable, so we could cache the `hashCode` computation. It's likely it will be re-used, especially if the `Name` is from generated code.

**Comment by lukaseder** [comment]
Some benchmarks (will be updated):

**Before the improvement**

``` 
Benchmark                                      Mode  Cnt        Score       Error  Units
BenchmarkRecordMapper.intoImmutablePojoCache  thrpt    7  4119371.363 ± 20829.895  ops/s
```

**Comment by lukaseder** [comment]
**After improving the hashing**

```
Benchmark                                      Mode  Cnt        Score       Error  Units
BenchmarkRecordMapper.intoImmutablePojoCache  thrpt    7  5162498.578 ± 73091.354  ops/s
```

**After improving the hashing *and* caching it**

```
Benchmark                                      Mode  Cnt        Score       Error  Units
BenchmarkRecordMapper.intoImmutablePojoCache  thrpt    7  5444287.268 ± 47624.944  ops/s
```

So, both improvements are totally worth it, given that this `hashCode` computation is a bottleneck for heavy mappings.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.17.0, 3.16.7 (https://github.com/jOOQ/jOOQ/issues/13500), and 3.15.11 (https://github.com/jOOQ/jOOQ/issues/13501)
